homeassistant/components/mqtt/sensor.py

```diff
@@ -196,6 +198,18 @@ def _update_state(msg):
                     self._state,
                     variables=variables,
                 )
+
+            if payload is not None and self.device_class in (
+                DEVICE_CLASS_DATE,
+                DEVICE_CLASS_TIMESTAMP,
+            ):
+                if (payload := dt_util.parse_datetime(payload)) is None:
```

Comment:
It's a bit meh to change the type of `payload`, but maybe that can be fixed in a future PR typing MQTT.